text/0001-workspace-consistent-versions.md

```diff
@@ -0,0 +1,572 @@
+
+Walking through an example of the confusing behavior:
+
+```json5
+// package.json
+{
+  "pnpm": {
+    "workspaceConsistent": {
+      "versions": {
+        "react": "^17.0.0",
+      }
+    }
+  }
+}
+```
+
+Assume the `react` specifier `^17.0.0` resolved to `17.0.2`.
+
+```yaml
+# pnpm-lock.yaml
+workspaceConsistent:
+  react:
+    specifier: ^17.0.0
+    dependency: 17.0.2
+```
+
+Suppose semver ranges were indeed allowed on `workspace-consistent:`. If a package were to specify `workspace-consistent:^`:
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent:^",
+  },
+}
+```
+
+On publish this is expected to become:
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "^17.0.0",
+  },
+}
+```
+
+It would be surprising if this became `^17.0.2` (created by adding `^` to the resolved version) since that may be overly specific and leave authors with no way to relax that.
+
+On the other hand, suppose a package were to specify `workspace-consistent:*`:
+
+```json5
+// packages/foo/package.json
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "workspace-consistent:*",
+  },
+}
+```
+
+On publish, the least suprising behavior would be for this to become `17.0.2`. This is because `17.0.2` was the resolved version actually tested on in the monorepo.
+
+```json5
+{
+  "name": "@monorepo/foo",
+  "dependencies": {
+    "react": "17.0.2",
+  },
+}
+```
+
+This leads to a discrepancy between what `workspace-consistent:*` and `workspace-consistent:^` should append semver range types to on publish. (The original specifier vs the resolved version.)
+
+The semantics may become more clear if `workspaceConsistent` configured versions must be a concrete version, but this eliminates the ability to `pnpm update` `workspaceConsistent` values.
+
+### Lockfile `workspaceConsistent` block format
+
+The proposed format is:
+
+```yaml
+workspaceConsistent:
+  jest:
+    specifier: ^27.5.1
+    dependency: 27.5.1
+
+  react:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+
+  react-dom:
+    specifier: ^17.0.2
+    dependency: 17.0.2
+```
+
+An alternative format that represents the same information would be:
+
+```yaml
+workspaceConsistent:
+  specifiers:
+    jest: ^27.5.1
+    react: ^17.0.2
+    react-dom: ^17.0.2
+  dependencies:
+    jest: 27.5.1
+    react: 17.0.2
+    react-dom: 17.0.2
+```
+
+The alternative format better matches the existing shape of `importers`. However the proposed format is more resilient to merge conflicts. Separate commits changing the version of `react` and `jest` would merge conflict with the alternative format, while the proposed format would merge those changes cleanly.
```

Comment:
OK, so then wouldn't it be enough to change the format of the importers? Merge conflicts would be fixed without adding the complex workspaceConstraints logic to the lockfile.

Reply:
Looking at the scenarios mentioned:

1. Changes the version of a dependency line-adjacent to the upgraded dependency.
2. Add or removes a dependency line-adjacent to the upgraded dependency.
3. Adds a new declaration of the upgraded dependency.

You're right. I think a new format for importers would address 1 and 2. 3 isn't handled, but that may be good enough. Will think about this a bit more.

Reply:
@gluxon I think we can implement this for importers and solve a big chunk of the issues. It is not hard to implement, so we can do it as a breaking change to the lockfile format in pnpm v7.

I think workspace constraints are a bigger change. I don't know yet if it is a good idea to put so much additional logic to the lockfile. In any case, this will require a lot of time for review.

Maybe there could be alternative solutions to solve the issues where the version is influenced by peer deps.

Like instead of this:

```yaml
importers:
  project:
    dependencies:

      react:
        specifier: ^1.0.0
        version: 1.0.0_react-dom@1.0.0

      react-dom:
        specifier: ^1.0.0
        version: 1.0.0
```

we could do this

```yaml
importers:

  project:
    dependencies:

      react:
        specifier: ^1.0.0
        version: ['1.0.0', 'react-dom']

      react-dom:
        specifier: ^1.0.0
        version: 1.0.0
```

Reply:
The situation @dmichon-msft described is something I see happen about once a week on my team. It's a flavor of the "_3. Adds a new declaration of the upgraded dependency._" conflict type (mentioned above) that the changes to the `importers` section described in the above comment would not address. At that frequency, it's not the most pressing problem. I tend to get an alert that our CI is broken, run `pnpm install`, and force push any updated `package.json` and `pnpm-lock.yaml` files to `develop`.

<details>
<summary>Tangent on bors</summary>
Last year I was trying to convince my team to adopt [bors](https://bors.tech/) as another way to prevent any kind of `develop` branch breaks, but I think we'll just switch to the [GitHub merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/using-a-merge-queue) when it's available on GitHub Enterprise instead of asking everyone to learn how bors works only to switch away from it later.
</details>

> I think we can implement this for importers and solve a big chunk of the issues. 

I think you're right. This solves 1 and 2 (https://github.com/pnpm/rfcs/pull/1#discussion_r838156141), which is indeed a big chunk of the problem.

<details>
<summary>Another tangent on `pnpm-lock.yaml` and newlines:</summary>
In January I ended up writing a `space-lockfile` command that is set to run on `postinstall`. It simply goes through `pnpm-lock.yaml` and adds any newlines before/after lines that contain an element in a list of package names. (The package names in this list are the consistent versions in the repo.) This made `pnpm-lock.yaml` merge conflict much less and has worked well since it was introduced 2 months ago, but the readability of the file was sacrificed a bit. I'm not completely satisfied with my `space-lockfile` solution either. If pnpm had a hook for serializing/deserializing lockfiles, I would have implemented a minimal version of this RFC without version groups with it instead.
</details>

> It is not hard to implement, so we can do it as a breaking change to the lockfile format in pnpm v7.

I appreciate you being open to this. I am a bit worried making the `importers` sections 25% newlines might cause other unexpected problems though, and wouldn't want to risk pnpm v7's release since it seems like we're in the beta phase already. Is it possible for these formatting changes to have edge cases where commits that merged cleanly before no longer do with the new format? Even if that is possible, perhaps that's a non-concern if it decreases the probability of conflicts as a whole.

> I think workspace constraints are a bigger change. I don't know yet if it is a good idea to put so much additional logic to the lockfile. In any case, this will require a lot of time for review.

I was imagining the `workspace-consistent:` protocol changes to primarily happen during `pnpm-lock.yaml` serialization and deserialization. Everything else in pnpm would see the same `PnpmLockfile` type as it does now. (With a `workspaceConsistent` key that can be ignored.) **Does that reduce the scope/risk of this change?** Or is it not practical to implement in this manner? There might need to be some tweaking to the `peersSuffix` format when there's a hash involved for this to happen during (de)serialization, but it feels doable.

> Maybe there could be alternative solutions to solve the issues where the version is influenced by peer deps.

I like the direction this idea is heading. I think the `peersSuffix` portion (especially when it's a hash) is the only non-intuitive part of the lockfile. I was only able to understand the concept [because you explained it well here](https://pnpm.io/how-peers-are-resolved).

Reply:
> I think we can implement this for importers and solve a big chunk of the issues. It is not hard to implement, so we can do it as a breaking change to the lockfile format in pnpm v7.

@zkochan I'd actually like to experiment with this a bit. We proposed this format above:

```yaml
importers:

  project:
    dependencies:

      react:
        specifier: ^1.0.0
        version: ['1.0.0', 'react-dom']

      react-dom:
        specifier: ^1.0.0
        version: 1.0.0
```

But I think we can just do:

```yaml
importers:

  project:
    dependencies:
      react:
        specifier: ^1.0.0
        version: ['1.0.0', 'react-dom']
      react-dom:
        specifier: ^1.0.0
        version: 1.0.0
```

The newlines aren't required since `react:` and `react-dom:` are on their own lines and not changing as much as `specifier` or `version`. Git can use those lines to determine where diff regions apply rather than the newlines.

@zkochan When I start experimenting, would it be preferable to localize this change to the lockfile serialization/deserialization, or update the `PnpmLockfile` shape itself? I lean towards the later despite the wider surface area.

Reply:
For now, I would localize the change to serialization/deserialization. Because we will probably hide the new format under a feature flag. Of course, when we'll be sure about the format, we will make it the default and refactor PnpmLockfile.

Reply:
Thanks for discussing! Filing an issue on myself to track: https://github.com/pnpm/pnpm/issues/4725